File: group_project/group_project/src/main/java/com/dikshanta/restaurant/management/system/group_project/service/RestaurantService.java
package com.dikshanta.restaurant.management.system.group_project.service;

import com.dikshanta.restaurant.management.system.group_project.configurations.SecurityAuditorAware;
import com.dikshanta.restaurant.management.system.group_project.dto.request.RestaurantCreateRequest;
import com.dikshanta.restaurant.management.system.group_project.dto.request.RestaurantStatusUpdateRequest;
import com.dikshanta.restaurant.management.system.group_project.dto.response.RestaurantCreateResponse;
import com.dikshanta.restaurant.management.system.group_project.enums.RestaurantStatus;
import com.dikshanta.restaurant.management.system.group_project.exceptions.RestaurantAlreadyExistsException;
import com.dikshanta.restaurant.management.system.group_project.exceptions.RestaurantDoesNotExistsException;
import com.dikshanta.restaurant.management.system.group_project.exceptions.UserDoesnotExistsException;
import com.dikshanta.restaurant.management.system.group_project.model.entities.Address;
import com.dikshanta.restaurant.management.system.group_project.model.entities.Restaurant;
import com.dikshanta.restaurant.management.system.group_project.model.entities.User;
import com.dikshanta.restaurant.management.system.group_project.repository.AddressRepository;
import com.dikshanta.restaurant.management.system.group_project.repository.RestaurantRepository;
import com.dikshanta.restaurant.management.system.group_project.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
public class RestaurantService {
    private final RestaurantRepository restaurantRepository;
    private final SecurityAuditorAware securityAuditorAware;
    private final UserRepository userRepository;
    private final AddressRepository addressRepository;

    public RestaurantCreateResponse createRestaurant(RestaurantCreateRequest request) {
        if (restaurantRepository.existsByName((request.getName()))) {
            throw new RestaurantAlreadyExistsException("Restaurant Already exists");

        }
        Long ownerId = securityAuditorAware.getCurrentAuditor()
                .orElseThrow(() -> new RuntimeException("User not authenticated"));
        User owner = userRepository.findById(ownerId).orElseThrow(() -> new UserDoesnotExistsException("Owner does not exists"));
        Address address = Address.builder()
                .province(request.getProvince())
                .district(request.getDistrict())
                .city(request.getCity())
                .street(request.getStreet())
                .build();
        Address restaurantAddress = addressRepository.save(address);
        Restaurant newRestaurant = Restaurant.builder()
                .name(request.getName())
                .description(request.getDescription())
                .owner(owner)
                .address(restaurantAddress)
                .imageUrl("/uploads/dikshanta")
                .status(RestaurantStatus.PENDING)
                .build();
        Restaurant savedRestaurant = restaurantRepository.save(newRestaurant);
        System.out.println(savedRestaurant);
        owner.getRestaurants().add(newRestaurant);
        return RestaurantCreateResponse.builder()
                .id(savedRestaurant.getId())
                .name(savedRestaurant.getName())
                .description(savedRestaurant.getDescription())
                .status(savedRestaurant.getStatus())
                .addressId(savedRestaurant.getAddress().getId())
                .ownerId(savedRestaurant.getOwner().getId())
                .imageUrl(savedRestaurant.getImageUrl())
                .build();


    }

    public void updateRestaurantStatus(RestaurantStatusUpdateRequest request) {
        Restaurant restaurant = restaurantRepository.findById(request.getRestaurantId()).orElseThrow(() -> new RestaurantDoesNotExistsException("Restaurant not found"));
        restaurant.setStatus(request.getRestaurantStatus());
        restaurantRepository.save(restaurant);
    }

    public List<Restaurant> getRestaurants() {

        return restaurantRepository.findAll();
    }

    public Restaurant getOwnRestaurant() {
        Long ownerId = securityAuditorAware.getCurrentAuditor()
                .orElseThrow(() -> new RuntimeException("User not authenticated"));
        return restaurantRepository.findByOwner(ownerId)
                .orElseThrow(() -> new RestaurantDoesNotExistsException("You don't have any restaurant registered"));
    }
}

File: group_project/frontend/src/services/restaurantService.ts
import api from './axios';
import { ApiResponse } from '../types/common';
import { RestaurantResponse, RestaurantCreateRequest } from '../types/restaurant';

interface RestaurantCreateResponse {
  id: number;
  name: string;
  description: string;
  status: string;
  imageUrl?: string;
}

export const restaurantService = {
  // GET /api/v1/restaurant/getRestaurant (gets own restaurant for logged-in RESTAURANT user)
  getOwnRestaurant: (): Promise<ApiResponse<RestaurantResponse>> =>
    api.get('/api/v1/restaurant/getRestaurant').then(r => r.data),

  // PUT /api/v1/restaurant/updateRestaurant/{id} (multipart)
  updateRestaurant: (id: number, request: RestaurantCreateRequest): Promise<any> => {
    const form = new FormData();
    form.append('name', request.name);
    form.append('description', request.description ?? '');
    if (request.province) form.append('province', request.province);
    if (request.district) form.append('district', request.district);
    if (request.city) form.append('city', request.city);
    if (request.street) form.append('street', request.street);
    if (request.multipartFile) form.append('multipartFile', request.multipartFile);
    return api.put(`/api/v1/restaurant/updateRestaurant/${id}`, form, {
      headers: { 'Content-Type': 'multipart/form-data' },
    }).then(r => r.data);
  },

  // DELETE /api/v1/restaurant/deleteRestaurant/{id}
  deleteRestaurant: (id: number): Promise<any> =>
    api.delete(`/api/v1/restaurant/deleteRestaurant/${id}`).then(r => r.data),
};
